src/work_pool.c

```diff
@@ -175,60 +141,93 @@ work_pool_thread(void *arg)
 	struct work_pool_thread *wpt = arg;
 	struct work_pool *pool = wpt->pool;
 	struct poolq_entry *have;
+	struct timespec ts;
+	int rc;
+	bool spawn;
 
-	atomic_inc_uint32_t(&pool->n_threads);
 	pthread_cond_init(&wpt->pqcond, NULL);
+	pthread_mutex_lock(&pool->pqh.qmutex);
+	TAILQ_INSERT_TAIL(&pool->wptqh, wpt, wptq);
+	pool->n_threads++;
 
 	do {
 		/* testing at top of loop allows pre-specification of work,
 		 * and thread termination after timeout with no work (below).
 		 */
 		if (wpt->work) {
-			if (pool->pqh.qcount > -pool->params.thrd_min
-			 && pool->n_threads < pool->params.thrd_max) {
+			spawn = pool->pqh.qcount < pool->params.thrd_min
+			      && pool->n_threads < pool->params.thrd_max;
+			pthread_mutex_unlock(&pool->pqh.qmutex);
+
+			if (spawn) {
 				/* busy, so dynamically add another thread */
 				(void)work_pool_spawn(pool);
 			}
-	
-			__warnx(TIRPC_DEBUG_FLAG_EVENT,
+
+			__warnx(TIRPC_DEBUG_FLAG_WORKER,
 				"%s() %s task %p",
 				__func__, pool->name, wpt->work);
 			wpt->work->wpt = wpt;
 			wpt->work->fun(wpt->work);
 			wpt->work = NULL;
+			pthread_mutex_lock(&pool->pqh.qmutex);
```

Comment:
Can initializing wpt be moved above the unlock above?  Since it's already in wptqh, it can be accessed while the lock is not held while it's uninitialized.

Reply:
OK. It's only used for diagnostics now anyway. Done.